### ISSUE TITLE
Partitioner stores fp8 copy of all weights between fwd and bwd, causing OOM

### ISSUE THREAD

**Issue by lw** [pytorch/pytorch#141881]
### 🐛 Describe the bug

We have [some code](https://github.com/facebookresearch/lingua/blob/main/lingua/float8.py) to convert linear layers to fp8. The weights are still stored in high precision, but we have an autograd.Function which converts them to fp8 in the forward and then again in the backward, in two slightly different ways. The autograd.Function does _not_ save the weight for the backward.

However, when we compile our code, the partitioner ends up choosing to fuse the two conversions into a single one, and save one of its results for the backward. Concretely, this means that the partitioner is choosing to store an additional copy of the entire model in fp8 between forward and backward! This amounts to multiple GBs of extra memory occupied, and is preventing training large models.

I don't have a strong opinion on how this should be fixed. I do _not_ think that the partitioner should be constrained to honor exactly what the autograd.Functions choose to keep/drop, but I do believe that the partitioner should take into account the amount of memory used by eager as an upper bound.

### Versions

Installed from the `pytorch-nightly` conda channel, v2.6.0.dev20241107, build py3.12_cuda12.4_cudnn9.1.0_0.

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu

**Comment by zou3519** [comment]
We've had multiple issues where users complain that torch.compile ends up using more memory when it doesn't faithfully reproduce an autograd.Function's "partitioning decision". Maybe an easy escape hatch is a decorator that will force torch.compile to respect it. But I'm hoping there's something more automatic the compiler can do.

**Comment by Chillee** [comment]
Hmm.... my feeling is that there's a bug/suboptimality in the partitioner here. In this case, the issue isn't even actually that it's recomputing the forwards pass when figuring out what to save. The problematic stuff is when choosing to save things that are neither in the forwards pass (i.e. required for output) nor the "backwards pass" (dependent on tangents).

**Comment by Chillee** [comment]
I actually have a hacky workaround for you - make the parts that you want to depend on the backwards pass have a "fake dependency" on `grad_out`, since that'll force it to execute in the backwards pass.

e.g.
```
    @staticmethod
    def backward(ctx, grad_out):
        a, b_t, amax_b = ctx.saved_tensors

        if ctx.a_requires_grad:
            b = b_t.t().contiguous() + (grad_out[0, 0] * 0)
            amax_grad_out = (grad_out.abs().amax(dim=-1, keepdim=True))
```

**Comment by lw** [comment]
Thanks, I'll try out the workaround!

As for the other comments, I think it's fine if the partitioner disregards the wishes of the autograd.Functions, as long as doing so makes the situation _better_...

**Comment by y-sq** [comment]
Hi @lw , can you try with this float8 training config? force_recompute_fp8_weight_in_bwd (https://github.com/pytorch/ao/blob/main/torchao/float8/config.py?#L225)
Context can be found in D63345959

**Comment by lw** [comment]
The workaround suggested by @Chillee seems indeed to work: the fake dependency will force the partitioner to move those nodes to the bwd.

**Comment by vkuzo** [comment]
@Chillee , thoughts on ETA and approach of longer term solutions in the partitioner without requiring a user workaround?

Unfortunately the workaround of adding a fake dependency via addition of 0 scalar (as above) doesn't work with FDSP2 with float8 all-gather, because the addition operation isn't defined for float8 dtypes in eager mode as it isn't supported by the underlying hardware.

**Comment by Chillee** [comment]
@vkuzo You can use any operator that induces a false dependency between the two - it doesn't have to be add. You could make your own custom operator that does nothing if you want :P

**Comment by vkuzo** [comment]
> @vkuzo You can use any operator that induces a false dependency between the two - it doesn't have to be add. You could make your own custom operator that does nothing if you want :P 

yah I get that, but I'm asking if we can fix this in the partitioner / when / how hard that would be? Adding a noop eager aten op for this to core just to work around is also not ideal.

**Comment by lw** [comment]
BTW, as I was adding support for tensor parallelism to our code I had to change a bit the workaround for it to get along with DTensors:
```
b_t = b_t + grad_out[:1, :, None].squeeze(0) * 0
```
(I used slice+squeeze because DTensors don't support select...)

**Comment by Chillee** [comment]
@vkuzo Yeah I'm still thinking about how to fix it properly in the partitioner. But I also don't think it's necessarily a bad idea to add some more explicit ways to override the partitioner.

**Comment by anijain2305** [comment]
@lw Do you have a workaround? Or is this still a high priority?

**Comment by vkuzo** [comment]
This is high priority from my POV, the workaround proposed above does work but it leads to difficult to maintain code.  Could someone chime in on ~O(effort) (days/weeks/months) and rough steps to fix this in the partitioner?

**Comment by lw** [comment]
The workaround we have is very high-maintenance (it broke when we tried to do tensor parallelism), and might prevent some performance optimizations (since we a fake dep which prevents the partitioner from doing anything). Thus it remain high-pri.

**Comment by eellison** [comment]
Would someone mind commenting a repro ?

**Comment by vkuzo** [comment]
I have a whole-model repro, in case that is useful:

1. check out torchtitan with https://github.com/pytorch/torchtitan/pull/808 applied (I'm going to land it very soon).
2. get a local editable install of torchao
3. run the following two experiments in torchtitan

```
# baseline (with workaround in this PR)
with-proxy CONFIG_FILE="./torchtitan/models/llama/train_configs/llama3_8b.toml" ./run_train.sh \
    --model.converters float8 \
    --training.compile \
    --float8.recipe_name rowwise

# experiment (remove the workaround, see peak memory usage increase)
# 1. hardcode out the workaround in torchao checkout: https://github.com/pytorch/ao/blob/1ab1b77ad744115c3fee62716e7d2083c57d80a1/torchao/float8/float8_linear.py#L166
# 2. then, rerun the same torchtitan command from above
with-proxy CONFIG_FILE="./torchtitan/models/llama/train_configs/llama3_8b.toml" ./run_train.sh \
    --model.converters float8 \
    --training.compile \
    --float8.recipe_name rowwise
```

On my machine (8xH100), I see peak memory usage increase from 1.04GiB to 1.05GiB on the debug torchtitan model with the ^ steps.

If someone wants to take a closer look at fixing this in a partitioner, happy to help get a smaller repro going.

**Comment by bdhirsh** [comment]
just to confirm - from talking to Luca, I think (if I understand correctly) the xformers repro is an example where the partitioner is behaving optimally for the subgraph it sees - it effectively doesn't know that FSDP is resizing a param to zero outside of the graph, and assumes that saving the parameter will be bad.

@vkuzo is your case similar? Horace pointed out higher up in this thread that there might be a case where the partitioner is just handling "unclaimed" nodes incorrectly, so I'm wondering if there is also separately a bug that can be fixed directly in the partitioner, without user annotations

**Comment by vkuzo** [comment]
@bdhirsh , yes, it's the same exact use case.
